src/spartan/upsnark.rs

```diff
@@ -615,4 +618,397 @@ impl<G: Group, EE: EvaluationEngineTrait<G>> PrecommittedSNARKTrait<G> for R1CSS
+    let claim_inner_joint =
+      self.claims_outer.0 + r * self.claims_outer.1 + r * r * self.claims_outer.2;
+
+    let (claim_inner_final, r_y) =
+      self
+        .sc_proof_inner
+        .verify(claim_inner_joint, num_rounds_y, 2, &mut transcript)?;
+
+
+    // verify claim_inner_final
+    // this should be log (num segments)
+    let N_PREFIX = (vk.num_vars_total.trailing_zeros() as usize - vk.num_steps.trailing_zeros() as usize) + 1;
+
+    let eval_Z = {
+      let eval_X = {
+        // constant term
+        let mut poly_X = vec![(0, 1.into())];
+        //remaining inputs
+        poly_X.extend(
+          (0..u.X.len())
+            .map(|i| (i + 1, u.X[i]))
+            .collect::<Vec<(usize, G::Scalar)>>(),
+        );
+        SparsePolynomial::new(usize::try_from(vk.num_vars_total.ilog2()).unwrap(), poly_X)
+          .evaluate(&r_y[1..])
+      };
+
+      // evaluate the segments of W
+      let r_y_witness = &r_y[1..N_PREFIX]; // skip the first as it's used to separate the inputs and the witness
+      let eval_W = (0..N_SEGMENTS).map(|i| {
+        let bin = format!("{:0width$b}", i, width = N_PREFIX-1); // write i in binary using N_PREFIX bits
+    
+        let product = bin.chars().enumerate().fold(G::Scalar::ONE, |acc, (j, bit)| {
+            acc * if bit == '0' {
+                G::Scalar::ONE - r_y_witness[j]
+            } else {
+                r_y_witness[j]
+            }
+        });
+    
+        product * self.eval_W[i]
+      }).sum::<G::Scalar>();
+      
+      (G::Scalar::ONE - r_y[0]) * eval_W + r_y[0] * eval_X 
+    };
+
+    // compute evaluations of R1CS matrices
+    let multi_evaluate_uniform = |M_vec: &[&[(usize, usize, G::Scalar)]],
+                          r_x: &[G::Scalar],
+                          r_y: &[G::Scalar], 
+                          num_steps: usize,|
+     -> Vec<G::Scalar> {
+      let evaluate_with_table_uniform =
+        |M: &[(usize, usize, G::Scalar)], T_x: &[G::Scalar], T_y: &[G::Scalar], num_steps: usize| -> G::Scalar {
+          (0..M.len())
+            .into_par_iter()
+            .map(|i| {
+              let (row, col, val) = M[i];
+              (0..num_steps).into_par_iter().map(|j| {
+                let row = row * num_steps + j;
+                let col = if col != vk.S_single.num_vars { col * num_steps + j } else { vk.num_vars_total }; 
+                let val = val * T_x[row] * T_y[col];
+                val
+              })
+              .sum::<G::Scalar>()
+            })
+            .sum()
+        };
+
+      let (T_x, T_y) = rayon::join(
+        || EqPolynomial::new(r_x.to_vec()).evals(),
+        || EqPolynomial::new(r_y.to_vec()).evals(),
+      );
+
+      (0..M_vec.len())
+        .into_par_iter()
+        .map(|i| evaluate_with_table_uniform(M_vec[i], &T_x, &T_y, num_steps))
+        .collect()
+    };
+    
+
+    let evals = multi_evaluate_uniform(&[&vk.S_single.A, &vk.S_single.B, &vk.S_single.C], &r_x, &r_y, vk.num_steps);
```

Comment:
This lambda does not get reused, let's remove the wrapper and add a comment instead.